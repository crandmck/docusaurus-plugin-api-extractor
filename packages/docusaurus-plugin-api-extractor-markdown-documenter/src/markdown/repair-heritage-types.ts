import visit from 'unist-util-visit';
import { Node, Data, Literal, Parent } from 'unist';

interface StrongNode extends Parent {
  type: 'strong';
}

/**
 * This function is responsible for fixing how the hertiage types
 * (Implements, Extends, References) are emitted as markdown with
 * bold tags instead of the markdown equivalent followed by a
 * markdown link. This seems to be problematic for some
 * implementations of markdown and causes the link not to render
 * properly.
 * @returns {}
 */
export const repairHeritageTypes = () => {
  return (tree: Node<Data>) => {
    visit<Parent>(tree, 'paragraph', (paragraphNode) => {
      const strong: StrongNode = {
        type: 'strong',
        children: [],
      };

      visit<Literal<string>>(paragraphNode, 'html', (htmlNode) => {
        const index = paragraphNode.children.indexOf(htmlNode);
        const textNode = paragraphNode.children[index + 1] as Literal<string>;

        if (
          htmlNode.value === '<b>' &&
          textNode.type === 'text' &&
          (textNode.value === 'Implements:' ||
            textNode.value === 'Extends:' ||
            textNode.value === 'References:')
        ) {
          const start = index + 1;

          let seenEndTag = false;
          paragraphNode.children.splice(index, 1, strong);

          for (
            let current = paragraphNode.children.length - 1;
            start <= current;
            current--
          ) {
            const newChild = paragraphNode.children[current] as Literal<string>;

            if (seenEndTag) {
              paragraphNode.children.splice(current, 1);
              strong.children.unshift(newChild);
            } else if (newChild.type === 'html' && newChild.value === '</b>') {
              seenEndTag = true;
              paragraphNode.children.splice(current, 1);
            }
          }
        }
      });
    });
  };
};
