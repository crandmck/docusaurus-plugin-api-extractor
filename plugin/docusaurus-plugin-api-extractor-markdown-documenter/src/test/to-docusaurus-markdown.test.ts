import { toDocusaurusMarkDown } from '../markdown/to-docusaurus-markdown';

it('can support union types in a table', () => {
  const incoming = `
  |  Property | Modifiers | Type | Description |
  |  --- | --- | --- | --- |
  |  [myProperty](./myClass.myProperty.md) |  | 'number' \\| 'string' | A property that has a union type. |
  `;

  const processedMarkdown = toDocusaurusMarkDown(incoming, 'myClass');
  expect(processedMarkdown).toMatchSnapshot();
});
