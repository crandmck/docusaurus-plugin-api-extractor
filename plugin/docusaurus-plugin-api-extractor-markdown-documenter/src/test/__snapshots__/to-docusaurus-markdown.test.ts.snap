// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can support union types in a table 1`] = `
"---
id: myClass
hide_title: true
custom_edit_url: null
---

| Property                              | Modifiers | Type                 | Description                       |
| ------------------------------------- | --------- | -------------------- | --------------------------------- |
| [myProperty](./myClass.myProperty.md) |           | 'number' \\\\| 'string' | A property that has a union type. |
"
`;
